Refactors MoreLikeThisQueryBuilder and Parser

alexksikes opened the pull request:
Relates to #10217

This PR is against the query-refactoring branch.

s1monw commented:
@alexksikes can you rebase this to latest - there are quite some changes here that are now gone

alexksikes commented:
@s1monw You can take a look now. Thank you.

colings86 commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -147,7 +204,8 @@ public Item doc(BytesReference doc) {
          */
         public Item doc(XContentBuilder doc) {
-            return this.doc(doc.bytes());
+            this.doc = doc != null ? doc.bytes() : null;
```
I don't think we should allow null here I think this should throw an exception if null is passed in

colings86 replied:
In fact, given that we require either doc or an index document id, this should really be a constructor and this setter should be removed. That way the user can only specify doc OR id and must specify one of them. We should add null checks in the constructors and throw IllegalArgumentException if nulls are passed in

colings86 replied:
Also, given that type and index are required if id is specified we should remove the setters for type, id and index and remove the zero-arg constructor for Item so the user is forced to correct initialise the object with either the index, type and id, or the index, type and doc

alexksikes replied:
There is a constructor for doc. However, we do have this method for bwc, so maybe we should keep it?

Concerning null index, or type, this is so that these could be initialized by the search context, and we only have this context when the query is actually being created.

colings86 replied:
We don't need to keep the constructors/methods for bwc. For the Java API we are able to break backwards compatibility here and we should if it makes the use of the API less trappy (as it does here).

alexksikes replied:
OK then if we are allowed to break bwc, then I much prefer to remove this method. That would simplify things from code and user perspective. A definite +1.

colings86 commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -326,6 +384,48 @@ public final String toString() {
+
+        @Override
+        public Item readFrom(StreamInput in) throws IOException {
+            Item item = new Item(in.readOptionalString(), in.readOptionalString(), in.readOptionalString());
```
This will need to change to read a boolean to tell it whether it needs to read a doc or Id so it can use the right constructor and doesn't trip up passing nulls into it

alexksikes replied:
Right now if you use a non null doc, then the id is ignored. If we remove the doc setter then yes we would have to do as you propose.

colings86 replied:
We should do this

alexksikes replied:
+1

colings86 commented on core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java:
```diff
@@ -0,0 +1,309 @@
+            queryBuilder.include(true);
+        }
+        if (randomBoolean()) {
+            queryBuilder.failOnUnsupportedField(false);
```
Why can't this and the one above be randomised?

alexksikes replied:
By default it is true, so then here we set it to false randomly.

colings86 replied:
We should set random boolean here too. We should be testing both the default (unset) and using the method will all allowed values (i.e. true and false)

alexksikes replied:
OK sounds good, however we do now have 2/3 of chance to get true, but I understand that in general we should be testing unset.

colings86 commented on core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java:
```diff
@@ -99,43 +71,11 @@
-import static org.hamcrest.Matchers.notNullValue;
-import static org.hamcrest.Matchers.nullValue;
-import static org.hamcrest.Matchers.sameInstance;
+import static org.hamcrest.Matchers.*;
```
Can you revert these changes as it's not a functional and will created conflicts to resolve when master is merged

alexksikes replied:
Yes we should do this upstream or revert this change. This is just running a optimize import on this class, which has way too many individual imports.

colings86 replied:
We shouldn't change classes just to clean up imports IMO. Unless there is an actual functional change to the class this just causes messy merges. Also, I don;t see why the number of individual imports makes any difference. Imports are hidden and automated by most IDEs and optimised out by the compile at compile time so have no affect on runtime (IIRC)

alexksikes replied:
Sure makes sense, I'll change that.

colings86 commented:
@alexksikes I left some comments. I also think someone else will need to quickly look before its merged as I have no experience of the MLT query

alexksikes commented:
@colings86 Thanks for the review. I'll address the comments on the next round of review.

cbuescher commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -316,6 +316,18 @@ public void writeStringArrayNullable(@Nullable String[] array) throws IOExceptio
 
+    /**
+     * Writes a string array, for nullable string, writes it as 0 (empty string).
+     */
```
Comment seems wrong, as far as I can see the difference to `writeStringArrayNullable` is that it writes boolean instead of int in case of null argument.

colings86 replied:
why is this method needed if we have `writeStringArrayNullable`?

cbuescher replied:
That one doesn't seem to differentiate between null and zero-length array, which we need if we e.g. test for equality of queries after serialization. At least that's what I think, @ale might have had other reasons.

alexksikes replied:
This exactly the reason. I first tried `writeStringArrayNullable`, but then you read back an empty array, not a null array.

colings86 replied:
I'm not a fan of having both of these methods. The difference between them is too subtle. I think we should instead replace `writeStringArrayNullable` with `writeOptionalStringArray` as the former is trappy and could easily lead to bugs. Serialisation code should always serialise and de-serialise to the same thing.

alexksikes replied:
+1 should this change be part of this PR? Looks like this is not used anywhere else.

colings86 replied:
If it's not used anywhere else (or even if its only used in a couple of places) then I would be +1 to doing it in this PR

alexksikes replied:
My bad this is actually used in a couple of places. I'll open an issue on this.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -46,10 +61,50 @@
      * A single item to be used for a {@link MoreLikeThisQueryBuilder}.
      */
-    public static final class Item implements ToXContent {
+    public static final class Item implements ToXContent, Writeable<Item> {
```
Given the amount of code in Item, I wonder if it would be good to move this to its own class. Also, lots of the complexity of this class seems to come from the basic two modes of operation (either as artificial doc or as indexed doc with index/type/id location). Maybe introducing two separate subclasses would help to seperate this. This shouldn't holp up this PR, maybe something for a follow-up, wdyt?

colings86 replied:
+1

alexksikes replied:
I agree this is what I first did in this pre-refactoring [PR](https://github.com/elastic/elasticsearch/pull/13372), but had to backtrack to keep bwc. Question: is query refactoring 3.0? How much are we allowed to break?

colings86 replied:
We should break anything we need to to make the Java API clean, easy to use, and less trappy. Given this is only going into 3.0 we should take this opportunity.

alexksikes replied:
I like that, thanks for the clarification.

cbuescher replied:
+1, only I don't think we need to hold this PR for this, this is big as it is already. I agree with Colin on cleaning up setters, constructors etc. for Item and get rid of as much ambiguity there as possible, but I think this is easier as a follow up

alexksikes replied:
Agreed I'll do the package change at last.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -78,29 +55,44 @@
 
+    public MoreLikeThisQueryParser() {
+
+    }
```
nit: this seems unused

alexksikes replied:
Sure I can remove. Note all parser seem to be having this empty unused constructor.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -116,37 +108,29 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
-                    }
-                } else if (parseContext.parseFieldMatcher().match(currentFieldName, Field.MINIMUM_SHOULD_MATCH)) {
-                    mltQuery.setMinimumShouldMatch(parser.text());
-                } else if ("analyzer".equals(currentFieldName)) {
```
Just for clarification, were those two clauses duplicates in the present version of the parser?

alexksikes replied:
I noticed that too. There are not duplicates in master but they are in the query refactoring branch. Probably something that happened during merge.

cbuescher commented:
@alexksikes I did a round of reviews, given the amount of complexity in constructing the final MLT query I didn't check the internals of the helper methods moved over from the parsers, maybe those can be simplified later but as far as separating the toQuery/fromXContent I think this is good. Maybe someone else should have final look after you adressed the last comments.

alexksikes commented:
@cbuescher @colings86 I addressed all the comments. Should I do the package change for the Item class as part of this PR too? Thanks again for the review.

s1monw commented on core/src/main/java/org/elasticsearch/index/VersionType.java:
```diff
@@ -304,4 +311,18 @@ public static VersionType fromValue(byte value) {
+
+    @Override
+    public VersionType readFrom(StreamInput in) throws IOException {
+        return VersionType.values()[in.readVInt()];
```
can we check the value we are reading here? an assertion would be ok

s1monw commented on core/src/main/java/org/elasticsearch/index/VersionType.java:
```diff
@@ -304,4 +311,18 @@ public static VersionType fromValue(byte value) {
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(this.ordinal());
```
`this` is unnecessary

s1monw commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -46,10 +61,50 @@
+
+    // document inputs
+    private final List<String> fields;
+    private List<String> likeTexts = new ArrayList<>();
```
I think the add and set methods for those lists here are too complex. I think we should only have setters and the setters should make an unmodifiable list out of the values as well as copy them.

s1monw commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -85,9 +142,12 @@ public Item() {
+        public Item(@Nullable String index, @Nullable String type, String id) {
             this.index = index;
             this.type = type;
+            if (id == null) {
```
can you move this up to fail before we assign

s1monw commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -98,10 +158,13 @@ public Item(String index, @Nullable String type, String id) {
             this.index = index;
             this.type = type;
-            this.doc(doc);
+            if (doc == null) {
```
can you move this up to fail before we assign

s1monw commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -517,6 +622,31 @@ public MoreLikeThisQueryBuilder maxWordLength(int maxWordLength) {
+        return this;
+    }
+
+    public MoreLikeThisQueryBuilder stopWords(List<String> stopWords) {
```
can we barf on null instead?

s1monw commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java:
```diff
@@ -679,4 +808,299 @@ private static void buildLikeField(XContentBuilder builder, String fieldName, Li
+        mltQuery.setMoreLikeFields(moreLikeFields.toArray(Strings.EMPTY_ARRAY));
+
+        // handle like texts
+        if (!likeTexts.isEmpty()) {
```
can you use `== false` instead of `!`

s1monw commented on core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java:
```diff
@@ -0,0 +1,291 @@
+
+public class MoreLikeThisQueryBuilderTests extends AbstractQueryTestCase<MoreLikeThisQueryBuilder> {
+
+    private static String[] randomFields;
```
these don't need to be static

s1monw commented:
I left some minor comments LGTM otherwise
